// @rork-ai/toolkit-sdk/lib/module/analytics.js
"use strict";

import { useGlobalSearchParams, usePathname } from "expo-router";
import PostHog, { PostHogProvider } from "posthog-react-native";
import { useEffect } from "react";
import { Platform } from "react-native";
import { isExpoGo } from "./utils.js";
import { jsx as _jsx } from "react/jsx-runtime";
const projectId = process.env.EXPO_PUBLIC_PROJECT_ID;
const teamId = process.env.EXPO_PUBLIC_TEAM_ID;
const isWeb = Platform.OS === "web";
const analyticsDisabled = !projectId || !teamId || isExpoGo() || isWeb;
const $groups = {
  project: projectId,
  team: teamId
};
let posthog = null;

// TODO

export function getPostHogClient() {
  if (analyticsDisabled) return null;
  if (!posthog) {
    posthog = new PostHog("phc_H2a6BKmqSmQJBp2iC6WcAFrRrZdeGX1c6YOiMh9qalF", {
      host: "https://toolkit.rork.com/relay-GnBZ/",
      captureAppLifecycleEvents: true,
      disableCompression: true
    });
    Object.entries($groups).forEach(([key, value]) => {
      posthog?.group(key, value, {});
    });
  }
  return posthog;
}
export function trackEvent(name, properties) {
  getPostHogClient()?.capture(name, {
    user_properties: properties,
    $groups
  });
}
export function RorkAnalyticsProvider({
  children
}) {
  const pathname = usePathname();
  const params = useGlobalSearchParams();
  useEffect(() => {
    getPostHogClient()?.screen(pathname, {
      user_params: params,
      $groups
    });
  }, [pathname, params]);
  const posthogClient = getPostHogClient();
  if (!posthogClient) return children;
  return /*#__PURE__*/_jsx(PostHogProvider, {
    client: posthogClient,
    autocapture: {
      captureTouches: false
    },
    children: children
  });
}
