// @rork-ai/toolkit-sdk/lib/module/integrations.js
"use strict";

import { useMutation, useQuery } from "@tanstack/react-query";
import * as WebBrowser from "expo-web-browser";
import { withBaseUrl } from "./api.js";
const EMPTY_ARRAY = [];
async function waitForConnection(id) {
  const response = await fetch(withBaseUrl(`/connection/wait/${id}`));
  return response.json();
}
export function useConnections() {
  const list = useQuery({
    queryKey: ["rork__connections"],
    queryFn: async () => {
      const response = await fetch(withBaseUrl("/connections/list"));
      return response.json();
    }
  });
  const initiate = useMutation({
    mutationFn: async toolkit => {
      const authUrl = withBaseUrl(`/connection/init/${toolkit}`);
      const response = await fetch(authUrl).then(res => res.json());
      const data = response;
      const resultPromise = WebBrowser.openAuthSessionAsync(data.redirectUrl, "demotoolkit://connection/callback");
      const result = await resultPromise;
      if (result.type === "success") {
        list.refetch();
      }
      return result;
    }
  });
  const disconnect = useMutation({
    mutationFn: async toolkit => {
      const response = await fetch(withBaseUrl(`/connection/disconnect/${toolkit}`));
      return response.json();
    }
  });
  return {
    currentConnections: list.data?.connections ?? EMPTY_ARRAY,
    refetchConnections: list.refetch,
    initiate,
    disconnect
  };
}
