// @rork-ai/toolkit-sdk/metro/index.js
const path = require("path");

/**
 * Metro configuration helper for Rork
 * Automatically wraps _layout.tsx with Rork providers
 * and redirects expo-haptics to web polyfill on web platform
 */

const libPath = path.resolve(__dirname, "../lib/module/");

function withRorkMetro(config) {
  const originalResolveRequest = config.resolver?.resolveRequest;

  return {
    ...config,
    transformer: {
      ...config.transformer,
      babelTransformerPath: require.resolve(
        "@rork-ai/toolkit-sdk/metro-transformer",
      ),
    },
    resolver: {
      ...config.resolver,
      extraNodeModules: {
        // whenever some lib does `require("assert")`
        // it will get this package
        assert: require.resolve("assert"),
        ...config.resolver?.extraNodeModules,
      },
      resolveRequest: (context, moduleName, platform) => {
        // Redirect expo-haptics to our web polyfill on web platform
        if (moduleName === "expo-haptics" && platform === "web") {
          return {
            filePath: path.resolve(libPath, "polyfills/haptics.web.js"),
            type: "sourceFile",
          };
        }

        if (moduleName === "expo-secure-store" && platform === "web") {
          return {
            filePath: path.resolve(libPath, "polyfills/secure-store.web.js"),
            type: "sourceFile",
          };
        }

        // Redirect react-native-maps to our web polyfill on web platform
        if (moduleName === "react-native-maps" && platform === "web") {
          return {
            filePath: path.resolve(libPath, "polyfills/maps.web.js"),
            type: "sourceFile",
          };
        }

        // Redirect react-native to our web polyfill on web platform
        // This replaces RefreshControl with our web implementation
        if (
          moduleName === "react-native-web/dist/exports/RefreshControl" &&
          platform === "web"
        ) {
          return {
            filePath: path.resolve(
              libPath,
              "polyfills/refresh-control-component.js",
            ),
            type: "sourceFile",
          };
        }

        // Redirect react-native to our web polyfill on web platform
        // This replaces RefreshControl with our web implementation
        if (
          moduleName === "react-native-web/dist/exports/Alert" &&
          platform === "web"
        ) {
          return {
            filePath: path.resolve(libPath, "polyfills/alert.web.js"),
            type: "sourceFile",
          };
        }

        if (originalResolveRequest) {
          return originalResolveRequest(context, moduleName, platform);
        }

        return context.resolveRequest(context, moduleName, platform);
      },
    },
    watcher: {
      watchman: {
        deferStates: ["rork-ai"],
      },
    },
  };
}

module.exports = { withRorkMetro };
