// @rork-ai/toolkit-sdk/metro/transformer.js
/**
 * Metro transformer that automatically wraps _layout.tsx with Rork providers
 * Uses simple string transformations to inject wrappers
 */

const fs = require("fs");
const path = require("path");

const upstreamTransformer = require("@expo/metro-config/babel-transformer");

/**
 * Detects the Expo SDK version from the installed expo package
 * This approach works with all package managers (npm, yarn, pnpm)
 */
function detectExpoVersion() {
  try {
    // Read version from installed expo package (works with pnpm, npm, yarn)
    const expoPackagePath = path.join(
      process.cwd(),
      "node_modules",
      "expo",
      "package.json",
    );
    if (fs.existsSync(expoPackagePath)) {
      const expoPackage = JSON.parse(fs.readFileSync(expoPackagePath, "utf8"));
      const version = expoPackage.version;
      if (version) {
        if (version.startsWith("54.")) return "54";
        if (version.startsWith("53.")) return "53";
      }
    }
  } catch {
    // Ignore parsing errors
  }
  return null;
}

/**
 * Checks if the file is the root _layout.tsx
 */
function isRootLayout(filename) {
  return (
    (filename.includes("app/_layout.tsx") ||
      filename.includes("app/_layout.js")) &&
    !filename.includes("app/(") // Exclude grouped layouts
  );
}

/**
 * Wraps the layout component with Rork providers
 */
function wrapLayoutWithProviders(src, environment) {
  // Find the default export function
  const defaultExportPattern = /export\s+default\s+function\s+(\w+)/;
  const match = src.match(defaultExportPattern);

  if (!match) {
    return src; // Can't transform, return as-is
  }

  const originalFunctionName = match[1];

  // Remove the default keyword from the original function
  let transformed = src.replace(
    defaultExportPattern,
    `function ${originalFunctionName}`,
  );

  const wrappers = [];

  // Add RorkDevWrapper for development environment
  if (environment === "development") {
    // Detect Expo SDK version - default to v53 if not detected
    const version = detectExpoVersion() || "53";
    const devImport =
      version === "54"
        ? "import { RorkDevWrapper } from '@rork-ai/toolkit-sdk/v54';"
        : "import { RorkDevWrapper } from '@rork-ai/toolkit-sdk/v53';";

    wrappers.push({
      import: devImport,
      name: "RorkDevWrapper",
    });
  }

  // Always add AnalyticsProvider
  wrappers.push({
    import: "import { RorkAnalyticsProvider } from '@rork-ai/toolkit-sdk';",
    name: "RorkAnalyticsProvider",
  });

  // Add wrapper imports at the top
  for (const wrapper of wrappers) {
    transformed = `${wrapper.import}\n${transformed}`;
  }

  // Create the wrapped component JSX
  let wrappedComponent = `<${originalFunctionName} />`;
  for (const wrapper of wrappers) {
    wrappedComponent = `<${wrapper.name}>${wrappedComponent}</${wrapper.name}>`;
  }

  // Add new default export that wraps the original function
  const newDefaultExport = `
export default function RorkRootLayoutWrapper() {
  return (
    ${wrappedComponent}
  );
}
`;

  transformed = transformed + "\n" + newDefaultExport;

  return transformed;
}

async function transform(props) {
  const environment =
    (props.options?.dev ?? process.env.NODE_ENV === "development")
      ? "development"
      : "production";

  if (isRootLayout(props.filename)) {
    const transformedSrc = wrapLayoutWithProviders(props.src, environment);
    return upstreamTransformer.transform({ ...props, src: transformedSrc });
  }

  return upstreamTransformer.transform(props);
}

module.exports = { transform };
